server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt

```diff
@@ -0,0 +1,233 @@
+    if (callExpression.getChildOfType<KtLambdaArgument>() != null) {
+        return
+    }
+
+    val resolvedCall = callExpression.getResolvedCall(file.compile)
+    val entries = resolvedCall?.valueArguments?.entries ?: return
+
+    val hints = entries.mapNotNull { (t, u) ->
+        val valueArg = u.arguments.firstOrNull()
+        if (valueArg != null && !valueArg.isNamed()) {
+            val label = getArgLabel(t.name, u)
+            valueArg.asElement().hintBuilder(InlayKind.ParameterHint, file, label)
+        } else null
+    }
+    acc.addAll(hints)
+}
+
+private fun getArgLabel(name: Name, arg: ResolvedValueArgument) =
+    (name).let {
+        when (arg) {
+            is VarargValueArgument -> "...$it"
+            else -> it.asString()
+        }
+    }
+
+private fun lambdaValueParamHints(
+    acc: MutableList<InlayHint>,
+    node: KtLambdaArgument,
+    file: CompiledFile,
+    config: InlayHintsConfiguration
+) {
+    if (!config.typeHints) return
+
+    val params = node.getLambdaExpression()!!.valueParameters
+
+    //hint should not be rendered when parameter is of type DestructuringDeclaration
+    //example: Map.forEach { (k,v) -> _ }
+    //lambda parameter (k,v) becomes (k :hint, v :hint) :hint <- outer hint isnt needed
+    params.singleOrNull()?.let {
+        if (it.destructuringDeclaration != null) return
+    }
+
+    val hints = params.mapNotNull {
+        it.hintBuilder(InlayKind.TypeHint, file)
+    }
+    acc.addAll(hints)
+}
+
+private fun chainedExpressionHints(
+    acc: MutableList<InlayHint>,
+    node: KtDotQualifiedExpression,
+    file: CompiledFile,
+    config: InlayHintsConfiguration
+) {
+    if (!config.chainedHints) return
+
+    ///chaining is defined as an expression whose next sibling tokens are newline and dot
+    val next = (node.nextSibling as? PsiWhiteSpace)
+    val nextSiblingElement = next?.nextSibling?.node?.elementType
+
+    if (nextSiblingElement != null && nextSiblingElement == DOT) {
+        val hints = node.getChildrenOfType<KtCallExpression>().mapNotNull {
+            it.hintBuilder(InlayKind.ChainingHint, file)
+        }
+        acc.addAll(hints)
+    }
+}
+
+private fun destructuringVarHints(
+    acc: MutableList<InlayHint>,
+    node: KtDestructuringDeclaration,
+    file: CompiledFile,
+    config: InlayHintsConfiguration
+) {
+    if (!config.typeHints) return
+
+    val hints = node.entries.mapNotNull {
+        it.hintBuilder(InlayKind.TypeHint, file)
+    }
+    acc.addAll(hints)
+}
+
+private fun declarationHint(
+    acc: MutableList<InlayHint>,
+    node: KtProperty,
+    file: CompiledFile,
+    config: InlayHintsConfiguration
+) {
+    if (!config.typeHints) return
+
+    //check decleration does not include type i.e. var t1: String
+    if (node.typeReference != null) return
+
+    val hint = node.hintBuilder(InlayKind.TypeHint, file) ?: return
+    acc.add(hint)
+}
+
+private fun functionHint(
```

Comment:
Why does this not have any tests? I know the code is small, but tests helps us not regress. It is also easier to extend the language server when I can TDDish it. (my own preference, but I think TDD is way more effective than the manual-test-loops many devs do).

Reply:
I did add a test for this:

https://github.com/fwcd/kotlin-language-server/blob/c58f60054f5ff5f9b3ceb14a7b833095ea7f42ae/server/src/test/kotlin/org/javacs/kt/InlayHintTest.kt#L77